examples/example_android_opengl3/main.cpp

```diff
@@ -16,6 +16,77 @@ static struct android_app*  g_App = NULL;
 static bool                 g_initialized = false;
 static char                 g_logTag[] = "ImguiExample";
 
+// Unfortunately, there is no way to show the on-screen input from native code.
+// Therefore, we call showSoftInput() of the main activity implemented in MainActivity.kt via JNI.
+static int showSoftInput()
+{
+    JavaVM *jVM = g_App->activity->vm;
+    JNIEnv *jEnv = NULL;
+
+    jint jniRet = jVM->GetEnv((void **)&jEnv, JNI_VERSION_1_6);
+    if (jniRet == JNI_ERR)
+        return -1;
+
+    jniRet = jVM->AttachCurrentThread(&jEnv, NULL);
+    if (jniRet != JNI_OK)
+        return -2;
+
+    jclass natActClazz = jEnv->GetObjectClass(g_App->activity->clazz);
+    if (natActClazz == NULL)
+        return -3;
+
+    jmethodID methodID = jEnv->GetMethodID(natActClazz, "showSoftInput", "()V");
+    if (methodID == NULL)
+        return -4;
+
+    jEnv->CallVoidMethod(g_App->activity->clazz, methodID);
+
+    jniRet = jVM->DetachCurrentThread();
+    if (jniRet != JNI_OK)
+        return -5;
+
+    return 0;
+}
+
+// Unfortunately, the native KeyEvent implementation has no getUnicodeChar() function.
```

Comment:
It can have if you try hard enough:

```cpp
uint32_t Android_AInputEvent_GetUnicodeChar(JNIEnv* env, struct AInputEvent* event)
{
    if (event == NULL || (AInputEvent_getType(event) != AINPUT_EVENT_TYPE_KEY))
        return 0;

    //KeyEvent(long downTime, long eventTime, int action, int code, int repeat, int metaState, int deviceId, int scancode, int flags, int source)
    int64_t downTime  = AKeyEvent_getDownTime(event);
    int64_t eventTime = AKeyEvent_getEventTime(event);
    int32_t action    = AKeyEvent_getAction(event);
    int32_t code      = AKeyEvent_getKeyCode(event);
    int32_t repeat    = AKeyEvent_getRepeatCount(event);
    int32_t metaState = AKeyEvent_getMetaState(event);
    int32_t deviceId  = AInputEvent_getDeviceId(event);
    int32_t scancode  = AKeyEvent_getScanCode(event);
    int32_t flags     = AKeyEvent_getFlags(event);
    int32_t source    = AInputEvent_getSource(event);

    // Create an instance of android.view.KeyEvent.
    jclass      keyEventClass             = env->FindClass("android/view/KeyEvent");
    jmethodID   keyEventClassConstructor  = env->GetMethodID(keyEventClass, "<init>", "(JJIIIIIIII)V");
    jobject     keyEventObject            = env->NewObject(keyEventClass, keyEventClassConstructor,
        downTime, eventTime, action, code, repeat, metaState, deviceId, scancode, flags, source);

    // Get unicode char.
    jmethodID   keyEventGetUnicodeChar    = env->GetMethodID(keyEventClass, "getUnicodeChar", "()I");
    int32_t unicodeChar = env->CallIntMethod(keyEventObject, keyEventGetUnicodeChar);

    // Release object
    env->DeleteLocalRef(keyEventObject);
    env->DeleteLocalRef(keyEventClass);

    return unicodeChar;
}
```

Reply:
Thanks for pointing that out. But still, this is not a native implementation for the KeyEvent.
I used this method before, but changed it in f8a0f03 to capturing KeyEvents in non-native code and let the characters be polled from the native side. This is less error prone, imo.

Reply:
Sometimes native functions in NDK are plainly broken. Showing soft keyboard for example was not working. Maybe now it is, but on android-9 it was not. Often they're often not on par with Android SDK. Like choreographer was missing from NDK for a few releases. Using SDK API to translate key event into Unicode codepoint is safe operation. One may not like dealing with JNI, but this is the nature of the beast.

Out of curiosity, can you elaborate more on why do you think pooling key events from Java is less error prone? It is bypassing native input queue.

Reply:
My approach of reconstructing the KeyEvent via JNI and calling `getUnicodeChar()` was only able to capture ASCII characters. I was doubtful about the reconstructed KeyEvent holding enough information to safely get Unicode characters. I figured that calling it on the non-native KeyEvent (Java/Kotlin side) is the best shot.

Concerning the input queue: You are right, it is based on the assumption, that all KeyEvents reach our Activity via `dispatchKeyEvent()`. I did not find any information if this is guaranteed. Your `Android_AInputEvent_GetUnicodeChar()` solution looks more promising than mine (I constructed the KeyEvent with event_type and key_code only). I will have a look.

About showing the soft keyboard: Yes, no reliable way to do it in native code. That was the reason to extend the NativeActivity in the first place.

Reply:
Maybe this isn't that big of a deal here. Key events would be useful when physical keyboard is connected to device. Text input will probably need to be implemented using IME.